#' read.wf reads in a .wf file (generated by comethylation.py) and stores the result as a WF instance.
#'
#' @param dir is the directory containing the .wf files. The default is NULL, which corresponds to the current working directory.
#' @param sampleName is the name of the sample and should be the prefix of the .wf files
#' @param n is an integer specifying the 'n' in n-tuple, i.e. the size of the n-tuples
#' @param methylationType is a character vector describing the what type of methylation loci are in the file: CpG, CHG or CHH.
#' @param chromosomes is a vector of chromosome names corresponding to the .wf files to be read by read.wf().
#' @param chromosomeLengths is a vector of the lengths of the 'chromosomes' argument; the order must be same as that for 'chromosomes'. 
#' @param mc.cores is the number of cores used in the mclapply backend (default: 1)

#' @return a WFComethylation instance of methylation loci n-tuples containing all .wf files as specified by the command line arguments
#' @keywords WF
#' @export
#' @examples
#' NULL
read.wf <- function(dir = NULL, sampleName, n, methylationType, chromosomes = 'all', chromosomeLengths = NULL, mc.cores = 1){
  ## Construct a list of chromosome names that are to have lag-correlations computed.
  message("Warnings involving 'Make sure to always combine/compare objects based on the same reference
  genome' can (probably) be ignored.")
  if(is.null(chromosomeLengths)){
    stop("The chromosome lengths must be supplied, with each element of 'chromosomeLengths' corresponding to the identical element of 'chromosomes'")
  }
  if(length(chromosomes) != length(chromosomeLengths))
    stop("'chromosomes' and 'chromosomeLengths' must be of the same length, with each element of 'chromosomeLengths' corresponding to the same element of 'chromosomes'")
  if(!is.integer(chromosomeLengths))
    stop("All elements of 'chromosomeLengths' must be integers")
  chromosome.list <- as.list(chromosomes)
  names(chromosome.list) <- chromosomes
  ## Read in files as a list
  ## Each element refers to a chromosome and each value is the corresponding .wf file for that chromosome
  wf_files <- list()
  filenames <- paste0(dir, sampleName, '_', names(chromosome.list), '_n=', n, '.wf')
  wf_list <- mcmapply(FUN = read.wfRaw, filenames, sampleName, n, methylationType, chromosomeLengths, mc.cores = mc.cores)
  combineList(wf_list)  
}

#' read.wfRaw is used internally by read.wf
#' @param wf_file is a .wf file
#' @param sampleName is the name of the sample
#' @param n is an integer specifying the 'n' in n-tuple, i.e. the size of the n-tuples
#' @param methylationType is a character vector describing the what type of methylation loci are in the file: CpG, CHG or CHH.
#' @param chromosomeLength is the length of the chromosome corresponding to data in the 'wf_file' (default: NA)
#' 
#' @return a WFComethylation instance of the wf_file
read.wfRaw <- function(wf_file, sampleName, n, methylationType, chromosomeLength = NA){
  ## Set up column headers by reading first line of file
  column_headers <- read.table(wf_file, header = F, sep = '\t', nrows = 1, stringsAsFactors = FALSE)
  ## Set up 'what' argument for scan()
  what0 <- replicate(length(column_headers), character(0))
  names(what0) <- column_headers
  char <- which(column_headers %in% "chr")
  what0[char] <- replicate(length(char), character(0))
  int <- which(!(column_headers) %in% "chr")
  what0[int] <- replicate(length(int), integer(0))
  ## Read in wf_file
  out <- scan(file = wf_file, what = what0, sep = "\t", quote = "", na.strings = "NA", quiet = TRUE, skip = 1)
  ## Construct WFComethylation instance
  gr <- GRanges(seqnames = out[['chr']], ranges = IRanges(start = out[['pos1']], end = out[[paste0('pos', n)]]))
  ## Recall: First column is chr, next n columns are positions, next 2^n columns are ntuples
  positions_index <- seq.int(from = 2, to = n + 1, by = 1)
  positions <- do.call(cbind, lapply(X = positions_index, FUN = function(x){out[[x]]}))
  colnames(positions) <- names(out)[positions_index]
  ntuples_index <- seq.int(from = n + 2, to = n + 1 + 2^n, by = 1)
  ntuples <- do.call(cbind, lapply(X = ntuples_index, FUN = function(x){out[[x]]}))
  colnames(ntuples) <- names(out)[ntuples_index]
  out_wf <- WFComethylation(gr = gr, positions = positions, ntuples = ntuples, sampleName = sampleName, methylationType = methylationType)
  seqlengths(out_wf) <- chromosomeLength
  return(out_wf)
}

